File: src/main/java/mc/fbturrets/util/VectorIterator.java
package mc.fbturrets.util;

import net.minecraft.util.math.Vec3d;

import java.util.Iterator;

public class VectorIterator implements Iterable<Vec3d> {
    private final Vec3d endVec;
    private Vec3d current = new Vec3d(0, 0, 0);
    private final Vec3d step;

    public VectorIterator(Vec3d vec, double scale) {
        endVec = vec;
        step = vec.normalize().multiply(scale);
    }

    @Override
    public Iterator<Vec3d> iterator() {
        return new Iterator<Vec3d>() {

            @Override
            public boolean hasNext() {
                return Double.compare(endVec.length(), current.length()) >= 0;
            }

            @Override
            public Vec3d next() {
                current = current.add(step);
                return current;
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }
}

File: src/main/kotlin/mc/fbturrets/util/VectorIterator.kt
package mc.fbturrets.util

import net.minecraft.util.math.Vec3d

class VectorIterator(private val endVec: Vec3d, scale: Double) : Iterable<Vec3d?> {
    private var current = Vec3d(0.0, 0.0, 0.0)
    private val step: Vec3d = endVec.normalize().multiply(scale)
    override fun iterator(): MutableIterator<Vec3d?> {
        return object : MutableIterator<Vec3d?> {
            override fun hasNext(): Boolean {
                return endVec.length().compareTo(current.length()) >= 0
            }

            override fun next(): Vec3d {
                current = current.add(step)
                return current
            }

            override fun remove() {
                throw UnsupportedOperationException()
            }
        }
    }

}